File: src/tester/network.ts
export class NetworkSelector {
  private config: any;
  public proxyUrl: string;
  public timeout: number;
  public testCount: number;
  public autoSelect: boolean;
  public forceMode: "direct" | "proxy";
  private apiKey: string;

  constructor(config: any) {
    this.config = config;
    this.proxyUrl = config.network?.proxy || "";
    this.timeout = config.network?.timeout || 10;
    this.testCount = config.network?.latency_test_count || 3;
    this.autoSelect = config.network?.auto_select !== false;
    this.forceMode = config.network?.force_mode || "direct";
    // Get the first valid API key
    this.apiKey = (config.api_keys || []).find((k: string) => k.trim())?.trim() || "";
  }

  private async measureLatency(proxy: string | null): Promise<number> {
    const latencies: number[] = [];
    const url = "https://integrate.api.nvidia.com/v1/models";

    for (let i = 0; i < this.testCount; i++) {
      try {
        const t0 = performance.now();
        const headers: Record<string, string> = {
          "Authorization": `Bearer ${this.apiKey}`,
        };

        const controller = new AbortController();
        const id = setTimeout(() => controller.abort(), this.timeout * 1000);

        const options: any = {
          method: "GET",
          headers,
          signal: controller.signal,
        };

        if (proxy) {
          options.proxy = proxy;
        }

        const resp = await fetch(url, options);
        clearTimeout(id);

        const elapsed = performance.now() - t0;
        if (resp.status < 500) {
          latencies.push(elapsed);
        }
      } catch (e: any) {
        // Quietly log to debug console if needed
      }
      await new Promise((resolve) => setTimeout(resolve, 300));
    }

    if (latencies.length === 0) {
      return Infinity;
    }
    return latencies.reduce((a, b) => a + b, 0) / latencies.length;
  }

  async selectBest(): Promise<[ "direct" | "proxy", number ]> {
    if (!this.autoSelect) {
      const mode = this.forceMode;
      console.log(`  ⚙️  强制模式：${mode}（已关闭自动选速）`);
      return [mode, 0.0];
    }

    console.log("  📡 测试直连延迟...");
    const directLatency = await this.measureLatency(null);
    console.log(
      directLatency !== Infinity
        ? `  直连平均延迟: ${directLatency.toFixed(0)}ms`
        : "  直连: ❌ 不可达"
    );

    let proxyLatency = Infinity;
    if (this.proxyUrl) {
      console.log(`  📡 测试代理延迟（${this.proxyUrl}）...`);
      proxyLatency = await this.measureLatency(this.proxyUrl);
      console.log(
        proxyLatency !== Infinity
          ? `  代理平均延迟: ${proxyLatency.toFixed(0)}ms`
          : "  代理: ❌ 不可达"
      );
    }

    if (directLatency === Infinity && proxyLatency === Infinity) {
      console.error("❌ 直连和代理均不可达，请检查网络或 API 密钥");
      throw new Error("网络不可达");
    }

    if (directLatency <= proxyLatency) {
      return ["direct", directLatency];
    } else {
      return ["proxy", proxyLatency];
    }
  }

  buildClientKwargs(mode: string): any {
    const kwargs: any = {};
    if (mode === "proxy" && this.proxyUrl) {
      kwargs.proxy = this.proxyUrl;
    }
    return kwargs;
  }
}

File: python_legacy/tester/network.py
# -*- coding: utf-8 -*-
"""网络测速模块：测试直连 vs 代理延迟，返回最优模式"""

import asyncio
import logging
import time
from typing import Tuple

import httpx

logger = logging.getLogger(__name__)

LATENCY_TEST_URL = "https://integrate.api.nvidia.com/v1/models"


class NetworkSelector:
    """自动测速选择最优网络路径"""

    def __init__(self, config: dict):
        self.config = config
        self.proxy_url = config["network"].get("proxy", "")
        self.timeout = config["network"].get("timeout", 10)
        self.test_count = config["network"].get("latency_test_count", 3)
        self.auto_select = config["network"].get("auto_select", True)
        self.force_mode = config["network"].get("force_mode", "direct")
        # 取第一个有效 API key 用于测速
        self.api_key = next(
            (k for k in config["api_keys"] if k.strip()), ""
        )

    async def _measure_latency(self, proxy: str | None) -> float:
        """测量指定代理配置的平均延迟（ms），失败返回 inf"""
        latencies = []
        headers = {"Authorization": f"Bearer {self.api_key}"}
        client_kwargs = dict(timeout=self.timeout, headers=headers, follow_redirects=True)
        if proxy:
            client_kwargs["proxy"] = proxy

        # 复用单个 client，避免每次循环重建连接池
        async with httpx.AsyncClient(**client_kwargs) as client:
            for _ in range(self.test_count):
                try:
                    t0 = time.perf_counter()
                    resp = await client.get(LATENCY_TEST_URL)
                    elapsed = (time.perf_counter() - t0) * 1000
                    if resp.status_code < 500:
                        latencies.append(elapsed)
                except Exception as e:
                    logger.debug(f"  延迟测试异常 proxy={proxy}: {e}")
                await asyncio.sleep(0.3)

        return sum(latencies) / len(latencies) if latencies else float("inf")


    async def select_best(self) -> Tuple[str, float]:
        """
        返回 (mode, avg_latency_ms)
        mode: "direct" | "proxy"
        """
        if not self.auto_select:
            mode = self.force_mode
            logger.info(f"  ⚙️  强制模式：{mode}（已关闭自动选速）")
            return mode, 0.0

        logger.info("  📡 测试直连延迟...")
        direct_latency = await self._measure_latency(None)
        logger.info(
            f"  直连平均延迟: {direct_latency:.0f}ms"
            if direct_latency != float("inf")
            else "  直连: ❌ 不可达"
        )

        proxy_latency = float("inf")
        if self.proxy_url:
            logger.info(f"  📡 测试代理延迟（{self.proxy_url}）...")
            proxy_latency = await self._measure_latency(self.proxy_url)
            logger.info(
                f"  代理平均延迟: {proxy_latency:.0f}ms"
                if proxy_latency != float("inf")
                else "  代理: ❌ 不可达"
            )

        if direct_latency == float("inf") and proxy_latency == float("inf"):
            logger.error("❌ 直连和代理均不可达，请检查网络或 API 密钥")
            raise RuntimeError("网络不可达")

        if direct_latency <= proxy_latency:
            return "direct", direct_latency
        else:
            return "proxy", proxy_latency

    def build_client_kwargs(self, mode: str) -> dict:
        """根据模式构建 httpx 客户端参数"""
        kwargs: dict = {
            "timeout": self.config["network"]["timeout"],
            "follow_redirects": True,
        }
        if mode == "proxy" and self.proxy_url:
            kwargs["proxy"] = self.proxy_url
        return kwargs
